# World Health Data Cleaning, Exploration, and Visualization
--------------------------
<p style='text-align: right;'> 01_Introduction.ipynb </p>
<p style='text-align: right;'> October 15 2020, David Diston </p>

### Introduction
------------------------------
My goal with this project is to demostrate the Data Scientist skills of Data Cleaning, EDA, and Visualization. As such, I will identify and explore unique and interesting observations in the data, but will not be drawing specific conclutions or making actionable suggestions. This project will be completed incrimentally, with each spcific step in the Cleaning/EDA/Visualization process being explored in it's own notebook.

The dataset for this project is titled: '<b>`Health Nutrition and Population Statistics `</b>` State of human health across the world`'. The data was originally collected by the `World Bank`; this version of the data has been aquired from `kaggle`. The dataset includes 345 indicators, such as immunization rates, malnutrition prevalence, and vitamin A supplementation rates across 258 countries/regions around the world. Data was collected on a yearly basis from 1960-2015.

### Initial Import, Exploration, and .CSV Creation by Country/Region
----------------------------------
In this introductory notebook, I will explore the overall shape of the data. This will include confirming the number of 'indicators' and countries/regions in the dataset, as well as looking into the formatting of values/columns, and the proportion of missing data.

I will also break apart the dataframe into individual csv's by countries/regions, and save these for potential use later.

In [1]:
import numpy as np
import pandas as pd

In [2]:
# the data has been downloaded and extracted into my working directory

datadf = pd.read_csv('data.csv')

In [3]:
# initial look at the shape/layout of the data

datadf.shape

(89010, 61)

In [4]:
datadf.sample(10)

Country Name Country Code  \
12147                        South Asia (IDA & IBRD)          TSA   
20312                                        Belarus          BLR   
3347   Europe & Central Asia (excluding high income)          ECA   
46886                                         Israel          ISR   
56624                                          Malta          MLT   
51824                                        Lebanon          LBN   
52537                                        Liberia          LBR   
76353                            St. Kitts and Nevis          KNA   
86063                                     Uzbekistan          UZB   
51225                                        Lao PDR          LAO   

                                          Indicator Name     Indicator Code  \
12147  Cause of death, by non-communicable diseases (...     SH.DTH.NCOM.ZS   
20312                                   Rural population        SP.RUR.TOTL   
3347   Population ages 50-54, male (% of male populat...  SP.POP.5054.MA.5Y   
46886           School enrollment, primary, male (% net)     SE.PRM.NENR.MA   
56624       Age population, age 16, female, interpolated  SP.POP.AG16.FE.IN   
51824       Children (ages 0-14) newly infected with HIV     SH.HIV.INCD.14   
52537  Exclusive breastfeeding (% of children under 6...     SH.STA.BFED.ZS   
76353                            Female population 40-44     SP.POP.4044.FE   
86063                              Male population 00-04     SP.POP.0004.MA   
51225                              Male population 35-39     SP.POP.3539.MA   

            1960       1961       1962       1963       1964       1965  ...  \
12147        NaN        NaN        NaN        NaN        NaN        NaN  ...   
20312  5541766.0  5498539.0  5456899.0  5414847.0  5370261.0  5327265.0  ...   
3347         NaN        NaN        NaN        NaN        NaN        NaN  ...   
46886        NaN        NaN        NaN        NaN        NaN        NaN  ...   
56624     2370.0     2906.0     3369.0     3832.0     4098.0     4063.0  ...   
51824        NaN        NaN        NaN        NaN        NaN        NaN  ...   
52537        NaN        NaN        NaN        NaN        NaN        NaN  ...   
76353        NaN        NaN        NaN        NaN        NaN        NaN  ...   
86063   672595.0   736537.0   801180.0   866265.0   924187.0   964658.0  ...   
51225    60381.0    61376.0    62395.0    63438.0    64510.0    65598.0  ...   

               2007          2008          2009          2010          2011  \
12147           NaN           NaN           NaN           NaN           NaN   
20312  2.555101e+06  2.503387e+06  2.455468e+06  2.409036e+06  2.363892e+06   
3347            NaN           NaN           NaN           NaN           NaN   
46886  9.679709e+01  9.680504e+01  9.659152e+01  9.651333e+01  9.620733e+01   
56624  2.582000e+03  2.695000e+03  2.855000e+03  2.957000e+03  2.919000e+03   
51824           NaN           NaN           NaN           NaN           NaN   
52537  2.900000e+01           NaN           NaN  3.400000e+01           NaN   
76353           NaN           NaN           NaN           NaN           NaN   
86063  1.354373e+06  1.397318e+06  1.444858e+06  1.509606e+06  1.593415e+06   
51225  1.493700e+05  1.538740e+05  1.583370e+05  1.625970e+05  1.673280e+05   

               2012          2013          2014       2015  Unnamed: 60  
12147  5.846883e+01           NaN           NaN        NaN          NaN  
20312  2.321803e+06  2.283483e+06  2.249652e+06  2219668.0          NaN  
3347            NaN           NaN           NaN        NaN          NaN  
46886  9.682038e+01  9.642369e+01  9.626073e+01        NaN          NaN  
56624  2.882000e+03  2.832000e+03  2.763000e+03     2682.0          NaN  
51824           NaN           NaN           NaN        NaN          NaN  
52537           NaN  5.520000e+01           NaN        NaN          NaN  
76353           NaN           NaN           NaN        NaN         

##### Observations
The data is arranged with years of the time series as columns. This is curious for time series data. Each row (observation) appears to be labeled by a dual 'index' of the `Country Name` and `Indicator Name`. Ideally I would like to change the layout of the data such that the `Indicator Names` are columns, and each row (observation) it identified by the `Year` and `Country Name`. This way, change in observation over time can more easily be made for specific `Indicator Names` grouped by `Country Name`. Prior to undertaking this transformation I will try to contextualize the density of data (in other words the amount of data vs. Null [NaN] observations).

In [5]:
datadf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 89010 entries, 0 to 89009
Data columns (total 61 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Country Name    89010 non-null  object 
 1   Country Code    89010 non-null  object 
 2   Indicator Name  89010 non-null  object 
 3   Indicator Code  89010 non-null  object 
 4   1960            35482 non-null  float64
 5   1961            35325 non-null  float64
 6   1962            35889 non-null  float64
 7   1963            35452 non-null  float64
 8   1964            35483 non-null  float64
 9   1965            35603 non-null  float64
 10  1966            35538 non-null  float64
 11  1967            36022 non-null  float64
 12  1968            35577 non-null  float64
 13  1969            35630 non-null  float64
 14  1970            36910 non-null  float64
 15  1971            37578 non-null  float64
 16  1972            38015 non-null  float64
 17  1973            37542 non-null 

##### Observations
There is a substantial amount of missing data in this dataset. The general trend seems to show a gradual decrease in missing values in more recent years culminating in a minimum in 2010 (57,099 of 89,010 possible observations present). However even in 2010 the proportion of non-null observations is substantial. 

However, due to the arrangement of the data it is difficult to immediately determine patterns in missing values (between `Country Names` and/or `Indicator Names`). Such information would be of much greater use than total missing observations by year. This conclusion gives further support for the rearrangement of the data.

In [6]:
# now many individual 'Indicators' are present in the dataset

len(datadf['Indicator Name'].unique())

345

In [7]:
# does each country have the same number of 'Indicators'

datadf['Country Name'].value_counts().unique()

array([345], dtype=int64)

##### Observations
Above I have confirmed that there are indeed 345 unique `Indicator Names` in the dataset. Of additional imporance, I have also determined that each individual `Country Name` in the dataset has the same number of `Indicator Names`. This is an important level of consistency in the dataset that may be important when manipulating this data in the future.

From the dataframe sample above, there appear to be non-numeric and redundant descriptive columns in the dataset that can be removed. Additionally, the `Unnamed: 60` column appears to be completely empty. Therefore it will be removed as well.

In [8]:
datadf.drop(['Country Code', 'Indicator Code', 'Unnamed: 60'], axis = 1, inplace = True)

In [9]:
# sanity check to ensure that the desired columns have been removed as intended

datadf.head()

Country Name                                     Indicator Name        1960  \
0   Arab World  % of females ages 15-49 having comprehensive c...         NaN   
1   Arab World  % of males ages 15-49 having comprehensive cor...         NaN   
2   Arab World  Adolescent fertility rate (births per 1,000 wo...  133.555013   
3   Arab World  Adults (ages 15+) and children (0-14 years) li...         NaN   
4   Arab World  Adults (ages 15+) and children (ages 0-14) new...         NaN   

         1961        1962        1963        1964        1965        1966  \
0         NaN         NaN         NaN         NaN         NaN         NaN   
1         NaN         NaN         NaN         NaN         NaN         NaN   
2  134.159119  134.857912  134.504576  134.105211  133.569626  132.675635   
3         NaN         NaN         NaN         NaN         NaN         NaN   
4         NaN         NaN         NaN         NaN         NaN         NaN   

         1967  ...       2006       2007       2008       2009       2010  \
0         NaN  ...        NaN        NaN        NaN        NaN        NaN   
1         NaN  ...        NaN        NaN        NaN        NaN        NaN   
2  131.665502  ...  50.329135  49.999851  49.887046  49.781207  49.672975   
3         NaN  ...        NaN        NaN        NaN        NaN        NaN   
4         NaN  ...        NaN        NaN        NaN        NaN        NaN   

        2011       2012       2013       2014  2015  
0        NaN        NaN        NaN        NaN   NaN  
1        NaN        NaN        NaN        NaN   NaN  
2  49.536047  49.383745  48.796558  48.196418   NaN  
3        NaN        NaN        NaN        NaN   NaN  
4        NaN        NaN        NaN        NaN   NaN  

[5 rows x 58 columns]

### Data Transposition
------------------------------
Below I will transpose the data such that each `Indicator Name` is a column in the dataset, and each `Year` represents a row (observation) of data corresponding to a particular `Country Name`. Years will be grouped together by `Country Name` in alphabetical order as they are in the original dataframe. 

To accomplish this process I will create a list of all `Country Names` over which I can iterate. I will then use this list to filter observations by each `Country Name`, transpose these county-specific observations, and then append these observations to a new transposed dataframe.

In [10]:
# first I will create a list of all unique `Country Names` in the dataset

country_list = datadf['Country Name'].unique().tolist()

In [11]:
# here I can also confirm the total number of countries in the dataset

len(country_list)

258

In [12]:
# I will begin by instantiating a new pandas dataframe

health = pd.DataFrame()

'''
By iterating over each 'Country Name' in the dataset I can filter the observations for this country from the 
original dataframe. I will assign these observations to a new temporary dataframe variable. 

I would like to export a .csv file of the data for each individual country. Since the data is already filtered
to a particular country, the `Country Name` is not necessary in the saved dataframe. Therefore I will remove it.

Prior to transposition I will make the `Indicator Names` the index, and also remove the index label by resetting
the axis name. This will result in clearer formatting after transposition.

After transposition, I will reset the index, and relabel the year column from 'index' to 'year'.

I will creat a name variable to use when saving the .csv file. Each file will be saved with the name of the country,
so for clarity in file naming, spaces and unwated characters will be removed and/or changed.

Now the temporary dataframe of the unique `Country Name` data can be saved as a .csv file.

Prior to appending the temporary dataframe to the new transposed dataframe, the `Country Name` column must be re-inserted.

The temporary dataframe is then appended to the current version of the newly created transposed dataframe.
'''

for i in country_list:
    df = datadf[datadf['Country Name'] == i].copy()
    df.drop('Country Name', axis = 1, inplace = True)
    df.set_index('Indicator Name', inplace = True)
    df.rename_axis(None, inplace = True)
    df = df.T.reset_index().rename(columns = {'index' : 'Year'})
    name = str(i)
    name = name.replace(" ", "_")
    name = name.replace(":", "_")
    name = name.replace("&", "and")
    df.to_csv(f'Raw_Data_By_Country/RAWDATA-{name}.csv', index = False)
    df.insert(0, 'Country Name', i)
    health = pd.concat([health, df])
    health.reset_index(drop = True, inplace = True)
    
health.to_csv('raw_health.csv', index = False)

In [13]:
# sanity check to make sure that the data transposition has occured as expected

health.sample(10)

Country Name  Year  \
10507                                             Norway  1995   
2223                                 Upper middle income  1999   
7688                                             Jamaica  1976   
1425   Middle East & North Africa (IDA & IBRD countries)  1985   
9825                                          Mozambique  1985   
9188                                               Malta  1964   
2283                                               World  2003   
6420                                               Ghana  1996   
10000                                              Nauru  1992   
6097                                              France  2009   

       % of females ages 15-49 having comprehensive correct knowledge about HIV (2 prevent ways and reject 3 misconceptions)  \
10507                                                NaN                                                                       
2223                                                 NaN                                                                       
7688                                                 NaN                                                                       
1425                                                 NaN                                                                       
9825                                                 NaN                                                                       
9188                                                 NaN                                                                       
2283                                                 NaN                                                                       
6420                                                 NaN                                                                       
10000                                                NaN                                                                       
6097                                                 NaN                                                                       

       % of males ages 15-49 having comprehensive correct knowledge about HIV (2 prevent ways and reject 3 misconceptions)  \
10507                                                NaN                                                                     
2223                                                 NaN                                                                     
7688                                                 NaN                                                                     
1425                                                 NaN                                                                     
9825                                                 NaN                                                                     
9188                                                 NaN                                                                     
2283                                                 NaN                                                                     
6420                                                 NaN                                                                     
10000                                                NaN                                                                     
6097                                                 NaN                                                                     

       Adolescent fertility rate (births per 1,000 women ages 15-19)  \
10507                                          13.990000               
2223                                           36.325252               
7688                                          152.880000               
1425                                           86.991430               
9825                                          179.687800               
9188                                           16.876200               
2283             

##### Observations
The resulting dataframe above now has each independent variable represented as a column, while each row (observation) can be identified by both the `County Name` and `Year`. There are several advantages to laying the data out in the form. Firstly, as will be seen below, it is easier to identify the proportion of missing data of each missing variable by country. Other advantages will be seen later in this project when potentially imputing missing data, and during visualization.

### Calculating the Proportion of Missing Data
-------------------------
It would have been simple to calculate the proportion of missing data for each year of observations, and even group this information by country, from the original dataframe. However, knowing the proportion of missing data for each year is somewhat more meaningless than the proportion of missing data for each independent variable. 

In the next step of this introductory analysis, I will determine the proportion of missing data, grouped by `Country Name`, for each independent variable. I will also calculate averages for each `Country Name`, and for each independent variable. This information will be useful in determining if a particular variable has enough data to be used in a future predictive model, or for meaningful data visualization.

In [14]:
# I will create a list of all the independent variables to iterate over

variable_list = datadf['Indicator Name'].unique().tolist()

# I will also create a list to be used as the column labels for my new data density dataframe

column_list = datadf['Indicator Name'].unique().tolist()
column_list.insert(0, 'Country Name')
column_list.append('Average Data Density')
column_list.append('Max Data Proportion')

In [15]:
# I start by instantiating my new dataframe, using the column list created above

data_density = pd.DataFrame(columns = column_list)

'''
I begin by iterating over a list of each Country Name, filtering the transposed data by country and assigning this
data to a new temporary dataframe. By dropping the 'Country Name', and 'Year' variables I am left with only numeric
columns. 

I will instantiate a new list that will contain the proportional data for each independent variable and country. I
will also instantiate two new variables to hold the cumulative proportion of all variables per country, and the maximum
proportion of all the variables for each country.

Using a nested loop I will iterate over each independent variable, summing the number of NaN occurances, and finding
the inverse proportion of that sum based on the total number of observations (number of years). Within this loop
I will append the data density proportion for each independent variable to the new_row list. Additionally I will
add each proportion to the cumulative sum, and check the current proportion against the current max proportion.

Following the completion of the nested loop I will add the 'Country Name' to the beginning of the new_row list, as well
as append the average proportion for all independent variables for that country, and the max proportion for that country.

    Note: the max proportion will establish a reference point for countries that may have been created
    or been disolved whithin the time series range of the data.
    
I will then append the new_row list to the data_density dataframe.

Next, I will create a list of the mean of each column (independent variable) in the dataframe. To account for the 
non-numeric 'Country Name' variable, I will insert a row-label at the begining of this list. I will then append
this list of averages as the final row of the new dataframe.

Finally I will save this new data_density dataframe as a .csv file to the working directory.
'''

for i in country_list:
    df = health[health['Country Name'] == i].copy()
    df.drop('Country Name', axis = 1, inplace = True)
    df.drop('Year', axis = 1, inplace = True)
    
    new_row = []
    total_prop = 0
    max_prop = 0
    
    for j in variable_list:
        data_proportion = round((1 - (df[j].isnull().sum() / len(df[j]))) * 100, 2)
        
        if data_proportion > max_prop:
            max_prop = data_proportion
        else:
            pass
        
        total_prop += data_proportion
        new_row.append(data_proportion)
        
    new_row.insert(0, i)
    avg_prop = round(total_prop / 345 , 2)
    new_row.append(avg_prop)
    new_row.append(max_prop)
    
    data_density.loc[len(data_density)] = new_row


mean_list = data_density.mean().to_list()
mean_list.insert(0, 'Average Data Density')
data_density.loc[len(data_density)] = mean_list

data_density.to_csv('data_density.csv', index = False)

In [16]:
# sanity check to make sure the new dataframe has been created correctly

data_density.sample(10)

Country Name  \
50                                        Aruba   
174                                     Morocco   
227                                   Swaziland   
134                                 Isle of Man   
256                                      Zambia   
22                          Lower middle income   
37   Sub-Saharan Africa (excluding high income)   
184                                       Niger   
27                                North America   
98                                      Ecuador   

     % of females ages 15-49 having comprehensive correct knowledge about HIV (2 prevent ways and reject 3 misconceptions)  \
50                                                0.00                                                                       
174                                               1.79                                                                       
227                                               3.57                                                                       
134                                               0.00                                                                       
256                                               5.36                                                                       
22                                                0.00                                                                       
37                                                0.00                                                                       
184                                               5.36                                                                       
27                                                0.00                                                                       
98                                                0.00                                                                       

     % of males ages 15-49 having comprehensive correct knowledge about HIV (2 prevent ways and reject 3 misconceptions)  \
50                                                0.00                                                                     
174                                               0.00                                                                     
227                                               3.57                                                                     
134                                               0.00                                                                     
256                                               5.36                                                                     
22                                                0.00                                                                     
37                                                0.00                                                                     
184                                               3.57                                                                     
27                                                0.00                                                                     
98                                                0.00                                                                     

     Adolescent fertility rate (births per 1,000 women ages 15-19)  \
50                                               98.21               
174                                              98.21               
227                                              98.21               
134                                               0.00               
256                                              98.21               
22                                               98.21               
37                                               98.21               
184                                              98.21               
27                                               98.21               
98                   

##### Observations
Now that I have created a dataframe that shows the proportion of real data present in the dataframe for each country and independent variable, I have a clearer idea of what information may potentially be gleaned from the data. For example, it is now easy to identify variables that have a very low data density, and remove them from consideration. Similarly, I could remove particular countries that have a low data density average. On a more specific scale, I can look at the availablility of data for certain variables between certain countries. This information will be very usful in future EDA and Data Visualization steps.

#### Future Steps
--------
I would like to limit my dataframe to variables that have a high degree of data density. I will then look to remove or impute any remaining missing values based on the remaining data in the dataframe. 